### PR TITLE
Update aframe asynchronously

### PR DESCRIPTION
**Pull request by RobertGemmaJr**
Sorry wanted to rework my branch structure a little. PR is the same:

- Renames `model.js` as `volume.js`
- Pass entire `models` array into `volume.js`
- Build initial mesh in `init` function
- Holds each individual model's data (model texture, material, transfer texture, etc) in the `modelsData` array
- Load each model's data asynchronously
- Error checking for invalid model and color map paths
- Cache models and color maps based on their paths (`path: data`)

I think moving forward we should refer to what's displayed as the `volume` which is made up of multiple models, each `model` being it's own image. Now that we're showing (working on) a blending of multiple models I gather it makes sense to refer to the component in aframe as something different which is why I renamed the file.

As of right now aframe just displays the first model without any blending. I wanted to get some eyes on this before I make too many changes so I'll work that into a different PR. `one-entity` is a staging branch (not sure if there's a technical term) that will get the blending changes too. It will all eventually be merged into `feat-multiple-models`. I left some extra console logs/comments/etc in the branch since we're not merging with main but that will all get deleted eventually. 

Will need to do a fair bit more deep comparison's with the models to determine what actually needs to be updated. Right now we're building a whole new mesh each time instead of just updating what's changed. Any ideas?

- closes issue #56

### CODE REVIEW

File: src/Aframe/volume.js [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+      new THREE.Mesh(
+        new THREE.BoxGeometry(1, 1, 1),
+        new THREE.ShaderMaterial(DEFAULT_MATERIAL)
+      )
+    );
+  },
+
+  update: function (oldData) {
+    // TODO: Only change this.modelsData based on difference between oldData and this.data
+    if (oldData.models !== this.data.models) {
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:68]: Yes, definitely. I think the deep comparison will become a separate PR though

File: src/Aframe/volume.js [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+  tick: function (time, timeDelta) {
+    const isVrModeActive = this.scene.is("vr-mode");
+    const mesh = this.getMesh();
+
+    // Position is controlled by controllerObject in VR
+    if (this.controllerObject && isVrModeActive) {
+      const grabObject =
+        this.controllerObject.el.getAttribute("buttons-check").grabObject;
+
+      if (this.grabbed && !grabObject) {
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:110]: aframe does have some ability to handle global state that we aren't using right now. Will probably be worth looking into for the various event listener functions

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:110]: Also renamed `clipPlane` as `gripDown` in `buttons-check.js`

File: src/Aframe/volume.js [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+          usedModels.set(model.path, texture);
+          resolve(texture);
+        },
+        () => {},
+        () => reject(new Error("Could not load the model at " + model.path))
+      );
+    });
+  },
+
+  async loadMaterial(model, texture) {
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:184]: Yes, I still need the wait for the function to return before moving to the next line where it's called (`update`)

File: src/Aframe/volume.js [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+      uniforms: uniforms,
+    });
+  },
+
+  async loadTransferTexture(colorMapPath, transferFunction) {
+    return await new Promise((resolve, reject) => {
+      const usedColorMaps = this.usedColorMaps;
+
+      // Load colorMap data (RGB)
+      new Promise((res, rej) => {
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:240]: I need to wait for `img.onload` to return but true, I'll make it a separate function

**Comment by broarr** [comment]
> Will need to do a fair bit more deep comparison's with the models to determine what actually needs to be updated. Right now we're building a whole new mesh each time instead of just updating what's changed. Any ideas?

I'm digging in trying to understand what you're asking here. There's an update function on the `Volume` a-frame component. In that function you mention that you want the model data to update only the difference between `oldData` and `this.data`. Am I in the right location? Where is the `update` method being called? And where is `this.data` initialized?

File: src/Aframe/volume.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Working locally on `aframe-deep-compare` branch, will merge with `one-entity` when it's ready

File: src/Aframe/volume.js [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+  },
+
+  /** HELPER FUNCTIONS */
+
+  getMesh: function () {
+    return this.el.getObject3D("mesh");
+  },
+
+  async loadTexture(model) {
+    return await new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Okay so, yes, they're unnecessary because it pauses on the function call (`await this.loadTexture(model.path)`). 

Interestingly, though, it doesn't work in the opposite direction. I can't just use async/await and remove the creation of the promise - I still have to create my own promise in order for `await` to have any effect. It seems like it's mostly semantic? By creating all the promises in a separate functions the async/await makes the `data.models.map(async (model) => {...}` code block much simpler. Otherwise I would need to chain several `.then()` blocks together.

TLDR, async/await doesn't have much benefit within a single function but rather when chaining multiple serialized promises together. It's there to "hide" `.then()` chains, not the `Promise`s as a whole. (I think? cc @broarr )

File: src/Aframe/volume.js [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+      Promise.all(
+        this.data.models.map(async (model) => {
+          // Load data
+          const name = await model.name;
+          const texture = await this.loadTexture(model);
+          const material = await this.loadMaterial(model, texture);
+          const transferTexture = await this.loadTransferTexture(
+            model.colorMap.path,
+            model.transferFunction
+          );
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: I reordered the calls to make a bit more sense, each element in the array runs async still. There might be some optimization left but I think it's about as good as we can make it without adding new complexity to the code. Pretty sure all there's left to do is wrap the `colorData` and `transferTexture` into a single awaited promise

**Comment by RobertGemmaJr** [comment]
> And where is `this.data` initialized?

`this.data` is to `aframe` as `props` is to `React`. The properties passed into the `volume` component (lines 45-48 in `AframeScene.jsx`) is wrapped inside the `this.data` object. 

> Where is the `update` method being called? 

`update` is a built in function that's called whenever the component's properties (`this.data`) changes. It's sort of like `prevProps` and `nextProps` in React, if that makes sense. Better explained [here](https://aframe.io/docs/1.3.0/introduction/writing-a-component.html#handling-property-updates). 


Right now I completely rebuild the volume whenever `update` is called. `if (oldData !== data)` will always be true, but eventually that check will become more granular. For example, if `oldData.models !== data.models` but `oldData.sliders === data.sliders` I only have to update parts of the volume's mesh relevant to `models` and vice versa. Plus it's all shallow comparisons right now but `data.models` is an array of objects. I've begun working on it in a local branch.

File: src/Aframe/volume.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (broarr) [reply src/Aframe/volume.js]: > async/await doesn't have much benefit within a single function

Disagree! Async/await allows your functions to use traditional try/catch semantics. This is huge! Dealing with exceptions with promises is a headache everyone always gets wrong. But maybe I'm misunderstanding your meaning here? What do you mean by a single function?

> I can't just use async/await and remove the creation of the promise - I still have to create my own promise in order for await to have any effect.

This is 💯% true! In this particular case you can't remove the promise and expect async/await to work, because the threejs api is callback based. In your `new Promise(...)` call you're doing the work to translate the threejs callback api to a promise/async api.

> It's there to "hide" .then() chains, not the Promises as a whole.

Also 💯% true! Async/await are syntactic sugar on top of promises. Everything you do with `.then()/.catch()` you can do with async/await and vice versa. Check out this article for a crude implementation of async functions using generators and promises. Reading similar code helped me have an "aha!" moment.

https://hallski.org/blog/async-await-with-generators/

I think that code might also give you a feel for what await "pausing" really means. It's pretty neat stuff! Before promises existed in the browser, babel would transform async code to something similar to this generator/promise. That's why everybody used babel/webpack back in the day 😉 

Alright, with that out of the way let's look at this block of code:

```js
async loadTexture(model) {
    return await new Promise((resolve, reject) => resolve(foo))
}
```

This code creates a new promise, waits until it's resolved, then returns a new promise that has already resolved. 

Now, here's @mcmcgrath13 's suggestion:

```js
loadTexture(model) {
  return new Promise((resolve, reject) => resolve(foo))
}
```

This code returns a new promise _and does not wait until it has been resolved_. 

The difference in these code blocks is subtle. If I understand promises correctly you're changing when the work of `loadTexture` gets enqueued into the microtask queue. This may or may not have an effect on your program (I'm not an expert in the microtask queue).

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Okay, yes! This is definitely making more sense. The issue Mary rightfully pointed out was that I was waiting at both the definition
```js
async loadTexture(model) {
    return await new Promise((resolve, reject) => resolve(foo))
}
```
and at the time of calling
```js
const texture = await loadTexture(model);
```

Ultimately I'm choosing to leave the await when the functions are called so it's more obvious that entire block of code is running asynchronously

```js
data.models.map(async (model) => {
          const { name, path, colorMap, transferFunction } = model;
          // Load texture from png
          const texture = usedModels.has(path)
            ? usedModels.get(path)
            : await this.loadTexture(model.path);

          // Load THREE DataTexture from color map's png and model.transferFunction
          const colorData = usedColorMaps.has(colorMap.path)
            ? usedColorMaps.get(colorMap.path)
            : await this.loadColorMap(colorMap.path);
          const transferTexture = await this.loadTransferTexture(
            colorData,
            transferFunction
          );

          const material = this.buildMaterial(model, texture, transferTexture);
          return {
            name: name,
            texture,
            material,
            transferTexture,
          };
        })
```

### Error Handling

Am I handling my errors okay? Right now I call 
```js
reject(new Error(...))
```
inside each individual promise and then
```js
Promise.all(...).then(...).catch((error) => {throw error})
```
This shows "Uncaught (in promise) [error]" on the console. The website loads but aframe is blank, I think this is the functionality I want. The errors come from invalid paths and the `reject()` error explains that clearly. Is this what's expected or should I look at restructuring the errors?

[the remark this answers is not in the document]
Comment (mcmcgrath13) [reply src/Aframe/volume.js]: > Okay, yes! This is definitely making more sense. The issue Mary rightfully pointed out was that I was waiting at both the definition
> 
> ```js
> async loadTexture(model) {
>     return await new Promise((resolve, reject) => resolve(foo))
> }
> ```
> 
> and at the time of calling
> 
> ```js
> const texture = await loadTexture(model);
> ```
> 
> Ultimately I'm choosing to leave the await when the functions are called so it's more obvious that entire block of code is running asynchronously

That's not quite what I was pointing out.  I was pointing out the async await on that function is redundant.  As you end up wrapping a promise in a promise (once desugared) with no in between code.  `awaits` (as far as I have grokked), don't cross function boundaries, so the await in the function definition is unrelated to the await at the point of calling.  At the point of calling you want to wait for the promise returned by the function to resolve, which with the async/await on the function definition is actually a promise resolving a resolved promise (so just a layer of indirection).  Does that help? Or maybe I'm just rambling (and possibly wrong) and not helping.  This is all pretty nuanced esotery, so not ultimately that important for this PR.

> ### Error Handling
> Am I handling my errors okay? Right now I call
> 
> ```js
> reject(new Error(...))
> ```
> 
> inside each individual promise and then
> 
> ```js
> Promise.all(...).then(...).catch((error) => {throw error})
> ```
> 
> This shows "Uncaught (in promise) [error]" on the console. The website loads but aframe is blank, I think this is the functionality I want. The errors come from invalid paths and the `reject()` error explains that clearly. Is this what's expected or should I look at restructuring the error handling?

Is there something more graceful you can do other than rethrowing the error?  It does throw a wrench into the machine, so does stop things like you want, but at that point, you could just not catch the error and get the same effect.  I think you could maybe `console.error` the message and not throw? (though not sure the implications for the rest of the running code at that point)

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Interesting. Nothing changes if I remove the `.catch((error) => {throw error})` block but if I do `.catch((error) => {console.error(error)})` it only displays my error message - the "uncaught (in promise)" is removed from the error message.

None of the three versions actually halts execution - just the creating of the volume in aframe. I think that's the preferred way to handle the errors though.

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: @mcmcgrath13  @broarr 

I added a try/catch block in the `Promise.all()` but I can't really figure out how to chain the errors together properly. Right now I `console.error` the original error (from the individual promise) and then throw a new error that includes the model's name. That is caught and thrown in the `.catch` block of `Promise.all`.

What I really want to do is throw the two errors separately but I can't seem to do that. If I do some thing like 
```js
throw new Error("Failed to load model '" + model.name + "'" + error)
```
it concatenates the message it doesn't throw two separate errors to the console. How can I chain multiple errors together as I go up the chain of promises?

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Or better yet, should I be checking the paths way higher in the react app? So I know the path is valid by the time I enter `volume.js`?

[the remark this answers is not in the document]
Comment (mcmcgrath13) [reply src/Aframe/volume.js]: I think you're looking for [`Promise.allSettled`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled).  But it also might just be easier to check the path upstream if you definitely can (but what if the data in the path is itself invalid?)

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Yeah, that's true. Right now I'm calling `reject()` inside of built in error callbacks to it makes sense to leave them there

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: I accidentally made my way to a pattern I really like. Instead of calling an error in the try/catch block I call `Promise.reject`. This displays the error on the console asynchronously (including the higher level error containing `model.name`) but doesn't halt execution. Then, because the try/catch block doesn't return anything, `this.buildMesh()` returns an error which the outermost `Promise.all().catch()` throws.

As far as I can tell this accomplishes the same task as `Promise.allSettled` without the need for the extra loop. The try/catch block will "fail" and display the errors for every model that fails to load. The catch block doesn't return anything so I know (at least) one modelData will be undefined in `this.buildMesh(modelsData)`. Right now I'm just looping over `modelsData` to force the error, I'll have to loop over the array to blend the models together anyways. 

Does that make sense? Let me know if it's a total antipattern or if I should always halt execution on the first error. They're displayed in order (promise-> try/catch -> Promise.all()) so I think it's the best way to go about everything.

File: src/Aframe/volume.js [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+      };
+      img.onerror = () => {
+        reject(new Error("Invalid colorMap path: " + colorMapPath));
+      };
+    });
+  },
+
+  // Create a THREE DataTexture from the RGB and A data
+  loadTransferTexture(colorData, transferFunction) {
+    return new Promise((resolve, reject) => {
```
Comment (mcmcgrath13) [inline src/Aframe/volume.js:244]: I don't see any callback/asyncness here - this can be a plain function I think

File: src/Aframe/volume.js [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+            return {
+              name,
+              texture,
+              material,
+              transferTexture,
+            };
+          } catch (error) {
+            // Display errors asynchronously
+            Promise.reject(error);
+            Promise.reject(new Error("Failed to load model '" + name + "'"));
```
Comment (mcmcgrath13) [inline src/Aframe/volume.js:115]: would it be better to reject once? I'm not sure how rejecting twice plays out

File: src/Aframe/volume.js [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+          }
+        })
+      )
+        .then((result) => this.buildMesh(result))
+        .catch((error) => {
+          // Halt execution (includes errors in this.buildMesh)
+          throw error;
```
Comment (mcmcgrath13) [inline src/Aframe/volume.js:122]: just to clarify my understanding, this approach allows all the data models to attempt to load, but then doesn't halt execution until after they've all loaded/not?

If so, that seems very reasonable to me

File: src/Aframe/volume.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: `throw` halts execution, even inside a try/catch, and is only displayed because of the `throw error` in `Promise.all().catch()`. `Promise.reject` displays the error no matter what but lets execution continue. So (as far as I can tell) calling `Promise.reject` inside the catch block doesn't tell `Promise.all` there's an error (and thus invoke it's `.catch()` block) whereas using `throw` will break execution out of the `Promise.all` and into `.catch()` before all of the promises inside `Promise.all` complete. That seems to be a behavior of `Promise.all` itself - it breaks at the first error. 

I'm not sure the semantics between the two and what goes on under the hood but that's the end result I've seen.

File: src/Aframe/volume.js (hunk 7 above)
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: I removed the promise and renamed as `buildTransferTexture()` because it's not actually loading anything

File: src/Aframe/volume.js (hunk 8 above)
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:115]: Having them separately will display the errors separately 

image

otherwise only the messages are combined

image

Personally I like separating them but let me know if you think I should combine them

File: src/Aframe/volume.js (hunk 9 above)
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:122]: Yup, exactly. Each `model` will either load properly or be undefined in the `results` array and execution halts inside `this.buildMesh()` if one of the model's is undefined

File: src/Aframe/volume.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (mcmcgrath13) [reply src/Aframe/volume.js]: Yeah, `Promise.all` rejects as soon as one of the tasks rejects, whereas `Promise.allSettled` lets everything finish executing.  While it would be another layer of logic, it might be cleaner/more idiomatic to swap this to an all settled usage then `.then(results => results.some(r => r.status == "rejected") && throw(new Error("failed"))`.

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Using `Promise.allSettled` hides all of the underlying errors through. The only thing that would display on the console is
image

Using `Promise.reject` inside the catch block will display the error but then all of the `results` have a "fulfilled" status and we're in the same boat.

### Error Cause
I did find this in the docs:
image

But chrome doesn't seem to support it. On Firefox it's what I was looking for
image

but chrome doesn't display it
image

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Also I still can't throw multiple errors so I think 
> my understanding of an `async/await` function is that a thrown error is turned into a promise rejection.

is untrue, at least for the current code block. I believe this might be because the errors are being thrown inside the try/catch so I will try removing them and seeing what happens.

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Yeah. The only place I can tell the user what model caused the error is inside the `data.models.map(async (model) => {` callback and the only way I can detect an error in the callback (in `this.loadTexture` and/or `this.loadColorMap`) is to use a try/catch block.

I suppose I'll have to loop through the `Promise.allSettled().then(results =>` array and display each error then?

[the remark this answers is not in the document]
Comment (mcmcgrath13) [reply src/Aframe/volume.js]: Yeah, throw does return from the function early.  I'm kind of surprised rejecting multiple times works - I'm curious what that desugars to 🤔

[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Yeah. It seems like `throw` cannot be used asynchronously and we want each loop of `data.models.map(async (model) => {` to be asynchronous (despite the inside of the loop being synchronous) which is why `Promise.reject` was preferable. `Promise.allSettled` forces each loop to finish which is why it works with `throw`.

That said, I now have the code working with `Promise.allSettled` and `throw`. We're looping through the promise array a few more times so (at least by eye) it takes a hair longer. I know we're only really expecting the library to used with a handful of models so maybe it's not the end of the world? 

I trust your judgement on whether `Promise.all` or `Promise.allSettled` is the cleaner solution

File: src/Aframe/volume.js [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+
+        if (errors.length) {
+          // Bubble errors up to AframeScene
+          document.dispatchEvent(
+            new CustomEvent("aframe-error", {
+              detail: errors,
+            })
+          );
+        } else {
+          // Blend model's into a single material and apply it to the model
```
Comment (mcmcgrath13) [inline src/Aframe/volume.js:123]: is this comment more of a todo?

File: src/Aframe/volume.js [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+          const modelsData = promises.map((p) => p.value);
+
+          // TODO: Blend all of the model's material into one
+          this.getMesh().material = modelsData[0].material;
```
Comment (mcmcgrath13) [inline src/Aframe/volume.js:127]: no more `buildMaterial` method call?

File: src/Aframe/volume.js [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+      );
+    } else {
+      uniforms.box_min.value = new Vector3(0, 0, 0);
+      uniforms.box_max.value = new Vector3(1, 1, 1);
+    }
+
+    // Create material
+    return new ShaderMaterial({
+      ...DEFAULT_MATERIAL,
+      uniforms: uniforms,
```
Comment (mcmcgrath13) [inline src/Aframe/volume.js:328]:
```suggestion
      uniforms,
```

File: src/components/AframeScene/AframeScene.jsx [hunk 13]
```diff
@@ -1,87 +1,72 @@
...
+      setErrors(e.detail);
+    };
 
-import { toAframeString } from "../../utils";
+    document.addEventListener("aframe-error", handler);
+
+    return () => document.removeEventListener("aframe-error", handler);
+  }, []);
+
+  console.log("Caught error", errors);
```
Comment (mcmcgrath13) [inline src/components/AframeScene/AframeScene.jsx:26]: remove console log?

File: src/Aframe/Shader.js [hunk 14]
```diff
@@ -0,0 +1,325 @@
+import { THREE } from "aframe";
+const { ShaderLib, Vector2, Vector3, Matrix4 } = THREE;
+
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/Shader.js:3]: Yes, the branch is made directly off of this one though (`aframe-blending`). That will also be merged into `one-entity` before everything is sent to main

File: src/Aframe/volume.js (hunk 10 above)
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:123]: Yeah. I changed that whole chunk to be in the `this.buildMesh()` function anyways so it'll show up there

File: src/Aframe/volume.js (hunk 11 above)
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:127]: `buildMesh` but yeah I think it's better to leave it as a separate function. Will revert

File: src/Aframe/volume.js (hunk 12 above)
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:328]: done

File: src/components/VolumeViewer/VolumeViewer.jsx [hunk 15]
```diff
@@ -30,7 +39,16 @@ function VolumeViewer({
...
     <Wrapper key={remountKey} className={className} style={style}>
-      <AframeScene models={models} sliders={sliders} />
+      <AframeScene
+        models={models}
+        position={position}
+        rotation={rotation}
+        scale={scale}
+        // position={position ?? DEFAULT_POSITION}
+        // rotation={rotation ?? DEFAULT_ROTATION}
+        // scale={scale ?? DEFAULT_SCALE}
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/components/VolumeViewer/VolumeViewer.jsx:49]: It's deleted on my end... this must be some old version? Hopefully it shows up with the new pushes I guess

File: example/src/App.js [hunk 16]
```diff
@@ -22,37 +22,32 @@ function App() {
   const [useTransferFunction, setUseTransferFunction] = useState(true);
   const [modelPath, setModelPath] = useState(salt);
 
+  thermal.path = "thermal/path";
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply example/src/App.js:25]: Yeah. I'll commented it out and leave a note since I don't really need it in the moment but will when I build out the error page

File: src/Aframe/volume.js (hunk 12 above)
Comment (mcmcgrath13) [reply src/Aframe/volume.js:328]: I don't see a change

Comment (RobertGemmaJr) [reply src/Aframe/volume.js]: Okay it should be committed now

File: src/components/AframeScene/AframeScene.jsx [hunk 17]
```diff
@@ -90,8 +75,29 @@ function AframeScene({ models, sliders }) {
         look-controls
         arcball-camera="initialPosition: 0 0 1;"
       />
+
+      {/* TODO: Replace with better error message */}
+      {errors.length && (
+        <Wrapper>
+          {errors.map((e) => (
+            <p key={e}>
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/components/AframeScene/AframeScene.jsx]: Sure thing! I'll make a nice error component to display at some point. Let me know if there's anything else I should add to it in the short time though #72

File: src/components/Controls/ClipControls.jsx [hunk 18]
```diff
@@ -16,7 +17,8 @@ function ClipControls({ sliders, setSliders }) {
   }
 
   return (
-    <Section title="Clip">
+    <Section title="">
+      <Title>Clip</Title>
```
Comment (mcmcgrath13) [inline src/components/Controls/ClipControls.jsx:21]: why the title move here?

File: src/components/Controls/ColorMapControls.jsx [hunk 19]
```diff
@@ -10,11 +10,13 @@ import {
...
 import "@reach/listbox/styles.css";
 
-import Section from "../Section";
+import Section from "./Section.jsx";
+import Title from "./Title.jsx";
 
 function ColorMapControls({ model, modelIdx, setModels }) {
   return (
     <Section title="Color Map">
+      <Title>Color Map</Title>
```
Comment (mcmcgrath13) [inline src/components/Controls/ColorMapControls.jsx:19]: same here - this one retained the section title, but also added a title

File: src/components/Controls/Section.jsx [hunk 20]
```diff
@@ -0,0 +1,11 @@
+import React from "react";
+import styled from "styled-components";
+
+function Section({ className, children }) {
+  return <Wrapper className={className}>{children}</Wrapper>;
+}
```
Comment (mcmcgrath13) [inline src/components/Controls/Section.jsx:6]: it looks like section no longer takes a title prop, remove those props from the places it's used?

File: src/components/Controls/ClipControls.jsx (hunk 18 above)
Comment (RobertGemmaJr) [reply src/components/Controls/ClipControls.jsx:21]: Whoops. Little artifact from merging in the `<EnabledControls>` branch. I separated the Section and Title into separate components so the prop is no longer necessary. Will delete.

File: src/components/Controls/Section.jsx (hunk 20 above)
Comment (RobertGemmaJr) [reply src/components/Controls/Section.jsx:6]: Yes, exactly. Done now but I'm running into the same push issues

Comment (RobertGemmaJr) [reply src/components/Controls/Section.jsx:6]: JK was just in the wrong local branch it should be all set now

File: src/utils/index.js [hunk 21]
```diff
@@ -73,25 +73,36 @@ function useModelsPropMemo(models) {
...
-  let str = "";
-  Object.entries(obj).forEach(([key, val]) => {
-    if (key === "colorMap") {
-      /* 
-        colorMap.path is either a png encoded string or the path to a png
+// Filter model properties needed from aframe
+function getAframeModels(models) {
+  const out = models.map((model) => {
+    // Pick only needed properties
+    model = pick(model, [
```
Comment (mcmcgrath13) [inline src/utils/index.js:80]: Is this purposefully overwriting model? might be cleaner to use a different variable name here to make sure there aren't any weird side effects

File: src/images/visible_off.svg [hunk 22]
```diff
@@ -0,0 +1,6 @@
+<!-- Material UI Visibility Off icon https://fonts.google.com/icons?selected=Material%20Icons%20Outlined%3Avisibility_off%3A -->
+<!-- Fill changed from #000000 to  -->
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/images/visible_off.svg:2]: I meant to change the color of them because they were oddly bold looking. I think just changing the size of them mitigated that a lot though

Will delete the comments

File: src/components/Controls/EnabledControls.jsx [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+          ...models.slice(modelIdx + 1),
+        ]);
+      }}
+    />
+  );
+}
+
+const VisibleImg = styled.img`
+  width: 18px;
+  height: 18px;
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/components/Controls/EnabledControls.jsx:28]: ✅

File: src/utils/index.js (hunk 21 above)
Comment (RobertGemmaJr) [reply src/utils/index.js]: changed to `aframeModel`
